controller/AdminAccessController.php

```diff
@@ -124,36 +111,96 @@ public function savePermissions()
             ), 500);
         }
 
-        //get resource
-        $clazz = new \core_kernel_classes_Class($resourceId);
         $resources = array($clazz);
         if($recursive){
             $resources = array_merge($resources, $clazz->getSubClasses(true));
             $resources = array_merge($resources, $clazz->getInstances(true));
         }
 
-        foreach($resources as $resource){
-            $permissions = $this->dataAccess->getDeltaPermissions($resource->getUri(),$privileges);
-            // add permissions
-            foreach ($permissions['add'] as $userId => $privilegeIds) {
-                if(count($privilegeIds) > 0){
-                    $this->dataAccess->addPermissions($userId, $resource->getUri(), $privilegeIds);
-                }
+        $success = true;
+        $code = 200;
+        $message = __('Permissions saved');
+        $processedResources = [];
+        foreach ($resources as $resource) {
+            $permissions = $this->dataAccess->getDeltaPermissions($resource->getUri(), $privileges);
+            try {
+                $this->removePermissions($permissions['remove'], $resource);
+            } catch (\common_exception_InconsistentData $e) {
+                $success = false;
+                $message = $e->getMessage();
+                break;
             }
-            // remove permissions
-            foreach ($permissions['remove'] as $userId => $privilegeIds) {
-                if(count($privilegeIds) > 0){
-                    $this->dataAccess->removePermissions($userId,$resource->getUri(),$privilegeIds);
+            $this->addPermissions($permissions['add'], $resource);
+            $processedResources[] = $resource;
+        }
+
+        if (!$success) {
+            $this->rollback($processedResources, $privileges);
+            $code = 500;
+        }
+
+        return $this->returnJson([
+            'success' => $success,
+            'message' => $message
+        ], $code);
+    }
+
+    /**
+     * @param array $permissions
+     * @param \core_kernel_classes_Resource $resource
+     * @throws \common_exception_Error
+     * @throws \common_exception_InconsistentData
+     */
+    private function removePermissions(array $permissions, \core_kernel_classes_Resource $resource)
+    {
+        $pp = $this->getPermissionProvider();

```

Comment:
$pp should be $permissionProvider. Don't use these short variable names please! ( $i is okay in for loop :) )